### PR TITLE
Add vertex index inversions

### PR DESCRIPTION
**Pull request by guyer**
Routines to determine cells and faces bordering vertices.
Based on https://stackoverflow.com/a/59686318/2019542

### CODE REVIEW

**Comment by tkphd** [comment]
Tests are failing.

```
----------------------------------------------------------------------
File "/root/project/fipy/tools/numerix.py", line 1033, in fipy.tools.numerix.invert_indices
Failed example:
    a = numpy.array([[0, 2], [1, 3], [0, 3], [3, 4]])
Exception raised:
    Traceback (most recent call last):
      File "/root/project/test-environment-27/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest fipy.tools.numerix.invert_indices[0]>", line 1, in <module>
        a = numpy.array([[0, 2], [1, 3], [0, 3], [3, 4]])
    NameError: name 'numpy' is not defined
----------------------------------------------------------------------
File "/root/project/fipy/tools/numerix.py", line 1034, in fipy.tools.numerix.invert_indices
Failed example:
    print(invert_indices(a, axis=0))
Exception raised:
    Traceback (most recent call last):
      File "/root/project/test-environment-27/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest fipy.tools.numerix.invert_indices[1]>", line 1, in <module>
        print(invert_indices(a, axis=0))
    NameError: name 'a' is not defined
```

File: fipy/tools/numerix.py [hunk 1]
```diff
@@ -1007,6 +1007,48 @@ def in1d(ar1, ar2, assume_unique=False):
...
+    return MA.masked_values(list(itertools.zip_longest(*x, fillvalue=-1)), -1)
+
+
+def invert_indices(arr, axis=-1):
+    """Invert an index array
+
+    Given an array of indices, return the locations in the array along
+    `axis` of each index.
+
+    >>> a = numpy.array([[0, 2], [1, 3], [0, 3], [3, 4]])
```
Comment (wd15) [inline fipy/tools/numerix.py]: Should be NUMERIX here, but maybe better to just remove NUMERIX altogether and use numpy

File: fipy/meshes/abstractMesh.py [hunk 2]
```diff
@@ -692,6 +692,80 @@ def _localOverlappingFaceIDs(self):
         """
         return self.topology._localOverlappingFaceIDs
 
+    @property
+    def _vertexCellIDs(self):
```
Comment (wd15) [inline fipy/meshes/abstractMesh.py]: This isn't actually used anywhere. Do you see a likely use for them?

Comment (wd15) [reply fipy/meshes/abstractMesh.py]: very nice docs BTW

File: fipy/tools/numerix.py [hunk 3]
```diff
@@ -1007,6 +1007,48 @@ def in1d(ar1, ar2, assume_unique=False):
...
+     [-- -- 2]]
+
+    After https://stackoverflow.com/a/59686318/2019542
+    """
+    import itertools
+
+    return MA.masked_values(list(itertools.zip_longest(*x, fillvalue=-1)), -1)
+
+
+def invert_indices(arr, axis=-1):
```
Comment (wd15) [inline fipy/tools/numerix.py]: Should this actually be in numpy proper?

**Comment by guyer** [comment]
@tkphd Tests should be fixed

File: fipy/tools/numerix.py (hunk 1 above)
Comment (guyer) [reply fipy/tools/numerix.py]: You know that as soon as we remove `NUMERIX` that they'll release some reason we'll want it back. It ain't hurtin' nuthin'.

File: fipy/meshes/abstractMesh.py (hunk 2 above)
Comment (guyer) [reply fipy/meshes/abstractMesh.py]: I needed `_vertexFaceIDs` for my parallel Gmsh writer and remembered that you'd done most of the work on StackOverflow. Once I implemented it, I realized I could radically simplify things but just using `_vertexCellIDs`, but figured I'd keep them both.

I'm making this a separate pull request just to keep things as atomic as possible. Gmsh I/O is going to be a monster code review as it is.

File: fipy/tools/numerix.py (hunk 3 above)
Comment (guyer) [reply fipy/tools/numerix.py]: Which? `lists_to_masked` or `invert_indices`? Or both?

Comment (guyer) [reply fipy/tools/numerix.py]: It would probably need to be in scipy due to the dependence on `scipy.sparse.coo_matrix`

Comment (wd15) [reply fipy/tools/numerix.py]: Oh yes, good point. Both could be useful I think. Not something to worry about now though.

File: fipy/meshes/abstractMesh.py (hunk 2 above)
Comment (wd15) [reply fipy/meshes/abstractMesh.py]: Oh, I just thought you were doing this for fun :-)

File: fipy/meshes/abstractMesh.py [hunk 4]
```diff
@@ -692,6 +692,80 @@ def _localOverlappingFaceIDs(self):
...
+        [[1 2 4 5]
+         [4 5 7 8]
+         [3 4 6 7]
+         [0 1 3 4]]
+        >>> print(m._vertexCellIDs) # doctest: +SERIAL
+        [[0 0 1 0 0 1 2 2 3]
+         [-- 1 -- 2 1 3 -- 3 --]
+         [-- -- -- -- 2 -- -- -- --]
+         [-- -- -- -- 3 -- -- -- --]]
+        """
```
Comment (tkphd) [inline fipy/meshes/abstractMesh.py]: Sorry, I do not understand how to get from the mesh to the array here; perhaps due to lack of familiarity with masked arrays?
The mapping of the vertices bounding each cell makes sense, though the order seems strange (vertices surrounding cell 1, 3, 2, 0, in counter-clockwise order).
@guyer, can you explain what the `_vertexCellIDs` means to me? Again, sorry... this looks like a simple accessor function, so I understand if the question is out of scope for this PR.

Comment (tkphd) [reply fipy/meshes/abstractMesh.py]: The [SO thread](https://stackoverflow.com/questions/59669831/method-for-cells-adjacent-connected-to-vertex-in-fipy/59686318#59686318) has `_vertexToCellIDs` as an `nCell`-length of length-4 lists, each containing the IDs cells with which the vertex is associated, up to the possible 4, padded by `-1`. This seems to make sense.
I'm unclear how that representation maps to this one.

Comment (tkphd) [reply fipy/meshes/abstractMesh.py]: @guyer/@wd15, nevermind -- I've got it.

**Pull request merged**